### ISSUE TITLE
Variability

### ISSUE THREAD

**Issue by hungvelo** [GoogleCloudPlatform/PerfKitBenchmarker#234]
We should have a metric to measure the repeatability/variation of the test results. I think performance consistency is as important as the performance itself.

**Comment by cmccoy** [comment]
I definitely agree on the importance. Do you mean across multiple runs of the same benchmark (e.g., running netperf on the same cloud, on the same VM type, in the same zone)?
Or within a single run (e.g. variability of TCP request/response latency within a run of netperf)?
We're making some progress on the second, but don't really have anything in place for the first built into PerfKitBenchmarker.

**Comment by hungvelo** [comment]
Yes, I meant across multiples runs of the same benchmark - how likely are multiple identical runs on the same cloud, same VM type, same OS, in the same zone give identical results.

**Comment by voellm** [comment]
This is one of the things PerfKitExplorer is boot at and that is showing
trends and variance.

For PKB how would you propose we show this?  Have a --iterations feature
and add a extra metrics for variance?

On Wed, Apr 29, 2015 at 4:19 PM, Hung <notifications@github.com> wrote:

> Yes, I meant across multiples runs of the same benchmark - how likely are
> multiple identical runs on the same cloud, same VM type, same OS, in the
> same zone give identical results.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/issues/234#issuecomment-97612448>
> .
>



-- 
Anthony F. Voellm (aka Tony)
Google Voice:  (650) 516-7382
<https://www.google.com/voice/b/0?pli=1#phones>
Blog: http://perfguy.blogspot.com/

**Comment by hungvelo** [comment]
Yes, maybe have 2 extra optional flags, one for variance metric (if desired) and one for number of iterations (with default of 5 iterations if not given?). Variance is defined as the average of the squared differences from the Mean.

**Comment by cmccoy** [comment]
I can't decide if this should be incorporated into PKB itself:

1. Defaulting to 5 iterations will make some benchmarks nearly untenable (HPCC and SPEC can already take > 1 day on some VM sizes).
2. If a benchmark is going to be repeated, I think the process should start from scratch, tearing down and re-create any resources used. Some variance may be specific to a VM instance (for example, a noisy neighbor); new VMs for each iteration will help capture this.

If others agree on the second point, each iteration becomes a complete PKB run.
We could add a tool (similar to the side-by-side tool) that automates repeatedly running PKB and calculates some summary statistics for each result.

**Comment by voellm** [comment]
The way I'd like to tackle this is to factor our the "optimize" functionality we have in the Redis and Mongodb benchmarks and turn that into standard functionality.  As part of the that we could include iterations.

More specifically what I have in mind:
* Each benchmark should declare its independent variables, rages of those values, and metrics.  This will allow code to step through the values, look at the metrics, and find the optimal point.
* Each measurement could be one or more iterations and for this we record the variance of the declared metrics.

To Conors points;
1. Agreed but that might be the users choice.  What we lack here is info for the user that tells them on average how long things take.  We've had this request before.
2. I agree the process should be complete for an iteration.  This will also make it easier to know what the output metrics are.
